blockprotocol/blockprotocol
# Support namespaced block names when publishing

## 🌟 What is the purpose of this PR?

This PR updates `blockprotocol` package and BP API to support namespaced names in the block metadata. This supplements work in https://github.com/blockprotocol/blockprotocol/pull/947 and https://github.com/hashintel/hash/pull/2024.




## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203358502199087/1203388867587773/f) _(internal)_

## 🔍 What does this change?

- `blockprotocol` package supports `@namespace/blockname` in addition to `blockname` in block metadata. When a namespace is defined, a suffixed API key is picked from `.blockprotocolrc` / env first.

- `apps/site/src/pages/api/blocks/publish.api.ts` supports `@namespace/blockname`; if there is a mismatch between the namespace and a token’s user, an error is thrown.

- `apps/site/src/pages/api/blocks/{create,update}.api.ts` are also updated for consistency, although we don't use these API routes

## 📜 Does this require a change to the docs?

Site docs and READMEs in block templates are updated

## 🐾 Next steps

- Publish packages to NPM, work on new GitHub action in HASH repo

## 🛡 What tests cover this?

- Linting
- Manual

## ❓ How to test this?

1. Checkout this branch
1. Start BP website locally
1. Create and copy API key for `@alice`
1. Checkout HASH repo
1. In HASH repo
   ```
   yarn install
   cd blocks/some-block
   yarn build
   BLOCK_PROTOCOL_SITE_HOST=http://localhost:3000 BLOCK_PROTOCOL_API_KEY=?? node ../../../path-to-blockprotocol-repo/libs/blockprotocol/cli.js  publish
   ```
   
1. Observe

   ```
   Unable to publish '@namespace/blockname' because the token belongs to 'alice' and not 'namespace'
   ```
   
1.  Open `blocks/some-block/dist/block-metadata.json` and replace `@namepsace` with `@alice`

1. Re-run publish command
1.  Observe
    ```
    Successfully published 'blockname'!
    Visit http://localhost:3000/@alice/blocks/blockname to see it on the Hub
    ```
